### PR TITLE
Allow Dedicated Servers to query map info from the Resource Center.

### LINKED ISSUE #3558: Maps not available on the server should not be selectable
but greyed out in the dedicated host scenario.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/459903-maps-not-available-on-the-server-should-not-be-selectable?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #10690: Re-read map files on dedicated servers without restarting it
I'm getting a few questions regarding the map sync on my servers if they are working or not, because when people join my servers and try to pick a map they've uploaded to resource center for greater than 5 minutes ago, they get "Map not found on server" - Even though it's there.

This seems to be because the maps are only loaded on server start (Same with client).

Is there any chance to put in a looping check of map files after the mod has been loaded? So the servers check the <mod>\maps every X minutes? (At least for dedicated servers, as these are usually unattended)

This way the servers don't need  to be restarted / play a game before they become available.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/30515982-re-read-map-files-on-dedicated-servers-without-restarting-it?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #10717: Clean up the filesystem code
The virtual filesystem (VFS) code is/was a giant ball of technical debt, and cleaning it up will win us a lot of perf and feature improvements.  This ticket tracks the previous and future progress on the issue.

- [x] Un-global the FileSystem class. [#9853]
- [x] Streamline the package interface, moving the various mix-format workarounds from engine code into the mix file parser. [#10525, #10530, #10535, #10554]
- [x] Replace direct-filesystem lookups with VFS lookups. [#10687, #10703, #10704, #10705, #10706, #10713]
- [x] Reference maps via their packages instead of paths [#10786]
- [x] Restore "All packages" listing in the asset browser [#10716]
- [ ] Separate Map and MapPreview, and only load Map when absolutely necessary:
  - [x] Remove Map references from MapPreview [#10697]
  - [x] Remove automatic map format updating [#10701]
  - [x] Load map metadata without initializing throwaway maps. [#10768]
  - [x] Load map previews without initializing throwaway maps.
  - [x] Use MapPreviews in the server code.
    - [ ] Support querying the resource API directly, removing the need for dedicated servers to sync maps. [#10717]
  - [x] Use MapPreviews in the lobby code.
- [x] Implement `IReadOnlyFileSystem` interface, and replace use of `Game.ModData.ModFiles` with a passed `IReadOnlyFileSystem` object. [#10765]
- [x] Split sequence initialization out of RuleSet, and store the SequenceProvider on the individual Map and defaults in a `DefaultSequences` dictionary (keyed by tileset) on ModData.
- [x] Remove the lazy loading in Map.
- [x] Add support for "oramod" packages. [[pchote/oramod](https://github.com/pchote/OpenRA/commits/oramod)]

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/30624381-clean-up-the-filesystem-code?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by pchote**
Dedicated servers will now query the map API in the same way that clients do.
This removes the need to manually rsync maps.

Closes #3558.
Closes #10690.
Closes #10717.

### CODE REVIEW

**Comment by pchote** [comment]
I recommend cherry-picking #10981 to make testing this less painful on a single machine.

**Comment by pchote** [comment]
http://resource.openra.net/maps/9515/ and http://resource.openra.net/maps/9514/ are available on the resource center to test with (delete them locally from the server and/or client first).

File: OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs [hunk 1]
```diff
@@ -338,64 +339,87 @@ public bool InterpretCommand(S server, Connection conn, Session.Client client, s
...
 							}
-							else if (c.Bot != null)
-								server.LobbyInfo.Clients.Remove(c);
-						}
 
-						// Validate if color is allowed and get an alternative it isn't
-						foreach (var c in server.LobbyInfo.Clients)
-							if (c.Slot == null || (c.Slot != null && !server.LobbyInfo.Slots[c.Slot].LockColor))
-								c.Color = c.PreferredColor = SanitizePlayerColor(server, c.Color, c.Index, conn);
+							// Validate if color is allowed and get an alternative it isn't
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs]: `if it isn't`?

File: OpenRA.Game/Map/MapPreview.cs [hunk 2]
```diff
@@ -305,11 +310,43 @@ public void UpdateRemoteSearch(MapStatus status, MiniYaml yaml)
 					GridType = r.map_grid_type;
 
 					Preview = new Bitmap(new MemoryStream(Convert.FromBase64String(r.minimap)));
+
+					var playersString = Encoding.UTF8.GetString(Convert.FromBase64String(r.players_block));
+					Players = new MapPlayers(MiniYaml.FromString(playersString));
```
Comment (obrakmann) [inline OpenRA.Game/Map/MapPreview.cs]: Wouldn't it make sense to have this in a try/catch block as well?

**Comment by obrakmann** [comment]
Seems to work fine, :+1:

File: OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs (hunk 1 above)
Comment (pchote) [inline OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs]: I don't understand your comment here...?  In any case, this is just moving existing code.

**Comment by pchote** [comment]
Fixed.

File: OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs [hunk 3]
```diff
@@ -814,8 +823,12 @@ void UpdateCurrentMap()
 					});
 				}).Start();
 			}
+			else if (Map.Status == MapStatus.DownloadAvailable)
+				LoadMapPreviewRules(Map);
 			else if (Game.Settings.Game.AllowDownloading)
-				modData.MapCache.QueryRemoteMapDetails(new[] { uid });
+			{
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs]: Braces around a single-line statement?

Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs]: These were left over from an earlier implementation that had multiple lines here.  Fixed.

File: OpenRA.Game/Map/MapPreview.cs [hunk 4]
```diff
@@ -272,47 +276,81 @@ bool EvaluateUserFriendliness(Dictionary<string, PlayerReference> players)
...
-			// Update on the main thread to ensure consistency
-			Game.RunAfterTick(() =>
+			if (status == MapStatus.DownloadAvailable)
 			{
-				if (status == MapStatus.DownloadAvailable)
+				// HACK/PERF: The following is very much not thread safe for code (mainly UI) that uses the
+				// properties. We can get away without locking because the outside code must always check the
+				// Status field, and knows not to access it unless it is Available or DownloadAvailable.
+				// This means that no other threads should be accessing these info fields until *after* the
+				// Status field is set to MapStatus.DownloadAvailable at the end of the method.
```
Comment (RoosterDragon) [inline OpenRA.Game/Map/MapPreview.cs]: This is no good - without some sort of synchronisation the UI thread may well never pick up these changes. One option might be to abstract all the data into a single volatile field and replace that field with a new reference you build here.

Comment (pchote) [inline OpenRA.Game/Map/MapPreview.cs]: The UI thread picks up the changes from the `Status` field changing.  That's what I was trying to explain with the comment.

Comment (pchote) [inline OpenRA.Game/Map/MapPreview.cs]: Using a single field with references into it sounds like a decent solution – certainly better than having to write a pile of getters with explicit locks.

Comment (pchote) [inline OpenRA.Game/Map/MapPreview.cs]: Done.

**Comment by pchote** [comment]
Updated and removed the original :+1: because the update changed a fundamental part of the PR.

**Comment by obrakmann** [comment]
:+1:

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/4877914d6f211c77d1a6c6ccc642bc3a2c2903d8...9608ab5fa8174518d46d0a07e10baa3d8fd333f7)

**Pull request merged**